File: CFD_Principles_and_Applications_Blazek/BookCode/C++/unstructured_2d/src/timeStep.cpp
/// @file timeStep.cpp
///
/// Computation of the time step.
///
//*****************************************************************************
//
//  (c) J. Blazek, CFD Consulting & Analysis, www.cfd-ca.de
//  Created February 15, 2014
//  Last modification: July 7, 2014
// 
//=============================================================================
//
//  This program is free software; you can redistribute it and/or
//  modify it under the terms of the GNU General Public License
//  as published by the Free Software Foundation; either version 2
//  of the License, or (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//
//*****************************************************************************

#include "timeDiscr.h"

/// Computes the maximum stable time step.
///
/// @param geometry    geometrical data
/// @param fluidProps  fluid properties
/// @param precond     low Mach-number preconditioning
/// @param order       order of the spatial discretization (1 or 2)
///
void TimeDiscr::TimeStep( const Geometry &geometry, const FluidProps &fluidProps,
                          const Precond &precond, int order )
{
  int  i;
  REAL sx, sy, ds, rrho, u, v, vc, cs, rhop, rhoT, hT, q2, theta, ra1g,
       a1, a4, a5, fmue, f1, f2, fac, dtv, cfac, lambdac, lambdav, tsmin;

  CONSVARS *cv    = fluidProps.cv;
  DEPVARS  *dv    = fluidProps.dv;
  NODE     *sproj = geometry.sproj;
  REAL     *vol   = geometry.vol;

  if (fluidProps.equsType == Equations::Euler)
  {
    // preconditioned Euler equations
    if (precond.switchedOn)
    {
      for (i=0; i<geometry.nndInt; i++)
      {
        sx       = sproj[i].x;
        sy       = sproj[i].y;
        ds       = sx + sy;
        u        = ABS(cv[i].xmom/cv[i].dens);
        v        = ABS(cv[i].ymom/cv[i].dens);
        rhop     =  cv[i].dens/dv[i].press;
        rhoT     = -cv[i].dens/dv[i].temp;
        hT       = dv[i].cpgas;
        q2       = u*u + v*v;
        theta    = precond.ComputeTheta( dv[i].gamma,dv[i].csoun,q2 );
        a1       = cv[i].dens*rhop*hT + rhoT;
        ra1g     = 1.0/(cv[i].dens*theta*hT+rhoT);
        a4       = a1*ra1g;
        a5       = cv[i].dens*hT*ra1g;
        vc       = (sx*u+sy*v)/ds;
        cs       = SQRT((vc*vc)*((a4-1.0)*(a4-1.0))+4.0*a5);
        tstep[i] = (2.0*vol[i])/((vc*(a4+1.0)+cs)*ds);
      }
    }
    // Euler equations
    else
    {
      for (i=0; i<geometry.nndInt; i++)
      {
        sx       = sproj[i].x;
        sy       = sproj[i].y;
        u        = ABS(cv[i].xmom/cv[i].dens);
        v        = ABS(cv[i].ymom/cv[i].dens);
        vc       = sx*u + sy*v;
        cs       = dv[i].csoun*(sx+sy);
        tstep[i] = vol[i]/(vc+cs);
      }
    }
  }
  else
  {
    VISCDEPVARS *dvLam = fluidProps.dvLam;
    if (order > 1)
      cfac = 1.0;
    else
      cfac = 2.0;

    // preconditioned Navier-Stokes equations (laminar)
    if (precond.switchedOn)
    {
      for (i=0; i<geometry.nndInt; i++)
      {
        sx       = sproj[i].x;
        sy       = sproj[i].y;
        ds       = sx + sy;
        rrho     = 1.0/cv[i].dens;
        u        = ABS(cv[i].xmom*rrho);
        v        = ABS(cv[i].ymom*rrho);
        rhop     =  cv[i].dens/dv[i].press;
        rhoT     = -cv[i].dens/dv[i].temp;
        hT       = dv[i].cpgas;
        q2       = u*u + v*v;
        theta    = precond.ComputeTheta( dv[i].gamma,dv[i].csoun,q2 );
        a1       = cv[i].dens*rhop*hT + rhoT;
        ra1g     = 1.0/(cv[i].dens*theta*hT+rhoT);
        a4       = a1*ra1g;
        a5       = cv[i].dens*hT*ra1g;
        vc       = (sx*u+sy*v)/ds;
        cs       = SQRT((vc*vc)*((a4-1.0)*(a4-1.0))+4.0*a5);
        fmue     = dvLam[i].mue/fluidProps.prlam;
        f1       = (4.0*rrho)/3.0;
        f2       = dv[i].gamma*rrho;
        fac      = MAX(f1,f2);
        dtv      = (fac*fmue)/vol[i];
        lambdac  = 0.5*((a4+1.0)*vc+cs)*ds;
        lambdav  = dtv*(sx*sx+sy*sy);
        tstep[i] = vol[i]/(lambdac+cfac*lambdav);
      }
    }
    // Navier-Stokes equations (laminar)
    else
    {
      for (i=0; i<geometry.nndInt; i++)
      {
        sx       = sproj[i].x;
        sy       = sproj[i].y;
        ds       = sx + sy;
        rrho     = 1.0/cv[i].dens;
        u        = ABS(cv[i].xmom*rrho);
        v        = ABS(cv[i].ymom*rrho);
        fmue     = dvLam[i].mue/fluidProps.prlam;
        f1       = (4.0*rrho)/3.0;
        f2       = dv[i].gamma*rrho;
        fac      = MAX(f1,f2);
        dtv      = (fac*fmue)/vol[i];
        vc       = sx*u + sy*v;
        cs       = dv[i].csoun*(sx+sy);
        lambdac  = vc + cs;
        lambdav  = dtv*(sx*sx+sy*sy);
        tstep[i] = vol[i]/(lambdac+cfac*lambdav);
      }
    }
  }

  // in case of global time-stepping - find min. time step in domain

  if (timeStepping == TimeStepping::Global)
  {
    tsmin = 1.0e+33;
    for (i=0; i<geometry.nndInt; i++)
    {
      tsmin = MIN(tsmin,tstep[i]);
    }
    for (i=0; i<geometry.nndInt; i++)
    {
      tstep[i] = tsmin;
    }
  }
}

File: CFD_Principles_and_Applications_Blazek/BookCode/Fortran/unstructured_2d/src/timeStep.f90
!> @file timeStep.f90
!!
!! Computation of the time step.
!
! *****************************************************************************
!
!  (c) J. Blazek, CFD Consulting & Analysis, www.cfd-ca.de
!  Created February 25, 2014
!  Last modification: June 3, 2014
!
! *****************************************************************************
!
!  This program is free software; you can redistribute it and/or
!  modify it under the terms of the GNU General Public License
!  as published by the Free Software Foundation; either version 2
!  of the License, or (at your option) any later version.
!
!  This program is distributed in the hope that it will be useful,
!  but WITHOUT ANY WARRANTY; without even the implied warranty of
!  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
!  GNU General Public License for more details.
!
!  You should have received a copy of the GNU General Public License
!  along with this program; if not, write to the Free Software
!  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
!
! *****************************************************************************

!> Computes maximum stable time step.
!!
subroutine TimeStep

  use ModDataTypes
  use ModGeometry
  use ModNumerics
  use ModPhysics
  use ModInterfaces, only : CompTheta
  implicit none

! local variables
  integer     :: i
  real(rtype) :: sx, sy, ds, rrho, u, v, vc, cs, rhop, rhoT, hT, q2, &
                 theta, ra1g, a1, a4, a5, fmue, f1, f2, fac, dtv, cfac, &
                 lambdac, lambdav, tsmin

! *****************************************************************************

  if (kequs == "E") then

! - preconditioned Euler equations

    if (kprecond == "Y") then
      do i=1,nndint
        sx       = sproj(1,i)
        sy       = sproj(2,i)
        ds       = sx + sy
        u        = Abs(cv(2,i)/cv(1,i))
        v        = Abs(cv(3,i)/cv(1,i))
        rhop     = cv(1,i)/dv(1,i)
        rhoT     = -cv(1,i)/dv(2,i)
        hT       = dv(5,i)
        q2       = u*u + v*v
        theta    = CompTheta( dv(4,i),dv(3,i),q2 )
        a1       = cv(1,i)*rhop*hT + rhoT
        ra1g     = 1.D0/(cv(1,i)*theta*hT+rhoT)
        a4       = a1*ra1g
        a5       = cv(1,i)*hT*ra1g
        vc       = (sx*u+sy*v)/ds
        cs       = Sqrt((vc*vc)*((a4-1.D0)*(a4-1.D0))+4.D0*a5)
        tstep(i) = (2.D0*vol(i))/((vc*(a4+1.D0)+cs)*ds)
      enddo

! - Euler equations

    else
      do i=1,nndint
        sx       = sproj(1,i)
        sy       = sproj(2,i)
        u        = Abs(cv(2,i)/cv(1,i))
        v        = Abs(cv(3,i)/cv(1,i))
        vc       = sx*u + sy*v
        cs       = dv(3,i)*(sx+sy)
        tstep(i) = vol(i)/(vc+cs)
      enddo
    endif

  else  ! kequs == "N"

    if (iorder > 1) then
      cfac = 1.D0
    else
      cfac = 2.D0
    endif

! - preconditioned Navier-Stokes equations (laminar)

    if (kprecond == "Y") then
      do i=1,nndint
        sx       = sproj(1,i)
        sy       = sproj(2,i)
        ds       = sx + sy
        rrho     = 1.D0/cv(1,i)
        u        = Abs(cv(2,i)*rrho)
        v        = Abs(cv(3,i)*rrho)
        rhop     = cv(1,i)/dv(1,i)
        rhoT     = -cv(1,i)/dv(2,i)
        hT       = dv(5,i)
        q2       = u*u + v*v
        theta    = CompTheta( dv(4,i),dv(3,i),q2 )
        a1       = cv(1,i)*rhop*hT + rhoT
        ra1g     = 1.D0/(cv(1,i)*theta*hT+rhoT)
        a4       = a1*ra1g
        a5       = cv(1,i)*hT*ra1g
        vc       = (sx*u+sy*v)/ds
        cs       = Sqrt((vc*vc)*((a4-1.D0)*(a4-1.D0))+4.D0*a5)
        fmue     = dv(6,i)/prlam
        f1       = (4.D0*rrho)/3.D0
        f2       = dv(4,i)*rrho
        fac      = Max(f1,f2)
        dtv      = (fac*fmue)/vol(i)
        lambdac  = 0.5D0*((a4+1.D0)*vc+cs)*ds
        lambdav  = dtv*(sx*sx+sy*sy)
        tstep(i) = vol(i)/(lambdac+cfac*lambdav)
      enddo

! - Navier-Stokes equations (laminar)

    else
      do i=1,nndint
        sx       = sproj(1,i)
        sy       = sproj(2,i)
        rrho     = 1.D0/cv(1,i)
        u        = Abs(cv(2,i)*rrho)
        v        = Abs(cv(3,i)*rrho)
        fmue     = dv(6,i)/prlam
        f1       = (4.D0*rrho)/3.D0
        f2       = dv(4,i)*rrho
        fac      = Max(f1,f2)
        dtv      = (fac*fmue)/vol(i)
        vc       = sx*u + sy*v
        cs       = dv(3,i)*(sx+sy)
        lambdac  = vc + cs
        lambdav  = dtv*(sx*sx+sy*sy)
        tstep(i) = vol(i)/(lambdac+cfac*lambdav)
      enddo
    endif  ! kprecond

  endif    ! kequs

! in case of global time-stepping - find min. time step in domain -------------

  if (ktimst == "G") then
    tsmin = 1.D+32
    do i=1,nndint
      tsmin = Min(tsmin,tstep(i))
    enddo
    do i=1,nndint
      tstep(i) = tsmin
    enddo
  endif

end subroutine TimeStep
